Add unique postcard errors + styleground fallback

bigkahuna443 opened the pull request:
First batch of unique postcard errors:
1. Tile with undefined ID
  a. This one is pretty beefy, but worth it imo. One of the most frustrating errors to locate, especially when it's a single tile.
2. Error applying Decal Registry property
3. Missing playback tutorial
4. Tileset XML issue
5. Generic XML syntax error

Also adds a fallback to the Misc atlas, which was the only thing missing for stylegrounds to not crash on missing textures.

May still add a few more errors later. There was also discussion about adding an option to select between ignoring a non-fatal error or displaying an error postcard, which seems like a good potential future goal.

Closes #312.

bigkahuna443 committed:
Add unique postcard errors + styleground fallback

First batch of unique postcard errors:
1. Tile with undefined ID
2. Error applying Decal Registry property
3. Missing playback tutorial
4. Tileset XML issue
5. Generic XML syntax error

Also adds a fallback to the Misc atlas, which was the only thing missing for stylegrounds to not crash on missing textures.

bigkahuna443 commented:
![badtileid_error](https://user-images.githubusercontent.com/13278973/187327761-583b527c-f5c5-49e8-aa77-19f10345bb81.PNG)
![decal_error](https://user-images.githubusercontent.com/13278973/187327763-1486a427-94e3-48f8-bc85-f865813e88fa.PNG)
![tilexml_error](https://user-images.githubusercontent.com/13278973/187327765-9ad65ffa-c71c-42b4-bbf6-bc82d5e50365.PNG)
![xml_error](https://user-images.githubusercontent.com/13278973/187327768-0f681a0c-ed9e-410c-872d-9b99e6fe8ca6.PNG)
![tutorial_error](https://user-images.githubusercontent.com/13278973/187327770-72ae2254-cc47-429f-b5ff-1fa865046684.PNG)

microlith57 commented on Celeste.Mod.mm/Patches/LevelLoader.cs:
```diff
@@ -18,6 +18,7 @@
 using MonoMod.Cil;
 using MonoMod.InlineRT;
 using MonoMod.Utils;
+using System.Text.RegularExpressions;
```
(nitpick) maybe move this up with the other `System` imports?

bigkahuna443 replied:
I'll do you one better and sort them lol

microlith57 commented on Celeste.Mod.mm/Patches/PlayerPlayback.cs:
```diff
@@ -0,0 +1,29 @@
+using Celeste.Mod;
+using Microsoft.Xna.Framework;
+using MonoMod;
+using System;
```
unused import
```suggestion
```

microlith57 commented on Celeste.Mod.mm/Patches/LevelLoader.cs:
```diff
@@ -190,8 +203,38 @@ public patch_LevelLoader(Session session, Vector2? startPosition = default)
-                Logger.Log(LogLevel.Warn, "LevelLoader", $"Failed loading map {sid}");
+                if (patch_LevelEnter.ErrorMessage == null) {
+                    if (e is KeyNotFoundException && e.StackTrace.Contains("TileHandler") && e.Data.Count != 0) {
+                        int x = (int)e.Data["x"];
```
would it be good to check if these keys are present?

bigkahuna443 replied:
`Exception.Data` is explicitly for user-defined purposes, and currently this is the only place Everest uses it (alternative was embedding data in the exception message + messy regex). If we expand exception handling we can add more checks but for now I think I prefer it cleaner
EDIT: I can add a length check at least

microlith57 reviewed (commented):
:+1:

bigkahuna443 committed:
Clean up usings, exception check

Kalobi commented on Celeste.Mod.mm/Patches/LevelLoader.cs:
```diff
@@ -37,7 +38,7 @@ public patch_LevelLoader(Session session, Vector2? startPosition = default)
         [MonoModConstructor]
         public void ctor(Session session, Vector2? startPosition = default) {
-            Logger.Log(LogLevel.Info, "LevelLoader", $"Loading map {session?.Area.GetSID() ?? "NULL"}");
+            Logger.Log(LogLevel.Info, "LevelLoader", $"Loading {session?.Area.GetSID() ?? "NULL"}");
```
Not sure about all the "map" removals in general, but this one in particular I'm not a fan of. I think making it clear that the thing being loaded is a map is helpful, especially since to someone who doesn't know what the LevelLoader does, but knows that Level is the internal name for rooms, might associate LevelLoader with rooms instead of maps.

bigkahuna443 replied:
Original load logs looked like this:
```
[Everest] [Info] [LevelLoader] Loading level flowerhouse/QM2/5_SawrfingCastle
(just crashes on LoadLevel error lmao)
(no log on loadlevel success)
[Everest] [Warn] [LoadLevel] Failed loading level 12 (SID: flowerhouse/QM2/5_SawrfingCastle)
```

Proposed logs look like this:
```
[Everest] [Info] [LevelLoader] Loading flowerhouse/QM2/5_SawrfingCastle
[Everest] [Warn] [LevelLoader] Failed loading flowerhouse/QM2/5_SawrfingCastle
[Everest] [Verbose] [LoadLevel] Loading room 00 of flowerhouse/QM2/5_SawrfingCastle
[Everest] [Warn] [LoadLevel] Failed loading room 00 of flowerhouse/QM2/5_SawrfingCastle
```

The "map" addition was from my prior postcard PR to try and differentiate things since the original logs were so ambiguous. With the changes though, I think that room vs. no room makes a good enough distinction. The postcard warning has always used just the SID as well.

Kalobi commented on Celeste.Mod.mm/Patches/LevelLoader.cs:
```diff
@@ -155,8 +157,19 @@ public patch_LevelLoader(Session session, Vector2? startPosition = default)
+                        string lineNum = Regex.Match(e.Message, @"Line (\d+)").Groups[1].Value;
+                        patch_LevelEnter.ErrorMessage = Dialog.Get("postcard_xmlerror").Replace("((path))", path).Replace("((line))", lineNum);
+                        Logger.Log(LogLevel.Warn, "LevelLoader", $"Failed parsing {path}");
+                    } else if (e.StackTrace.Contains("Autotiler")) {
```
Is this really the cleanest way to determine that the error comes from the Autotiler? Could you use something like `e.TargetSite`, or do these errors generally happen in a called method that's no longer in `Autotiler`?

bigkahuna443 replied:
I'll look into `TargetSite`, pretty much every other `Exception` "feature" is just user-defined data so I guess I skipped over that. Will try and get rid of all the `StackTrace` string checks too

Kalobi commented on Celeste.Mod.mm/Patches/Autotiler.cs:
```diff
@@ -241,11 +241,16 @@ public patch_Autotiler(string filename)
+
+            // Satisfies error handling for the orig_ method too.
+            if (!lookup.TryGetValue(tile, out patch_TerrainType terrainType)) {
+                throw new KeyNotFoundException { 
```
The exception still eventually gets logged, so I think it should probably also have a human-readable error message as before.

Kalobi commented on Celeste.Mod.mm/Patches/LevelLoader.cs:
```diff
@@ -190,8 +203,38 @@ public patch_LevelLoader(Session session, Vector2? startPosition = default)
-                patch_LevelEnter.ErrorMessage = Dialog.Get("postcard_levelloadfailed").Replace("((sid))", sid);
-                Logger.Log(LogLevel.Warn, "LevelLoader", $"Failed loading map {sid}");
+                if (patch_LevelEnter.ErrorMessage == null) {
+                    if (e is KeyNotFoundException && e.StackTrace.Contains("TileHandler") && e.Data.Count == 3) {
```
Here we could just add a `Source` tag or something to `Data` instead of checking the StackTrace, or if we set the error message, we could base the check off that.

Kalobi commented on Celeste.Mod.mm/Patches/LevelLoader.cs:
```diff
@@ -190,8 +203,38 @@ public patch_LevelLoader(Session session, Vector2? startPosition = default)
+                        }
+
+                        string type = "";
+                        if (e.StackTrace.Contains("SolidTiles")) {
```
What if it's a background tile error, but the stacktrace contains my new mod `SolidTilesStrackTracePolluter`? :3dpausecorn: I think these StackTrace.Contains checks are generally not ideal.

Kalobi commented on Celeste.Mod.mm/Patches/PlayerPlayback.cs:
```diff
@@ -0,0 +1,28 @@
+            } else {
+                patch_LevelEnter.ErrorMessage = Dialog.Get("postcard_missingtutorial").Replace("((tutorial))", tutorialName);
+                Logger.Log(LogLevel.Warn, "PlayerPlayback", $"Failed to load tutorial '{tutorialName}'");
+                throw new KeyNotFoundException();
```
This exception should have an error message.

bigkahuna443 replied:
I think specific exception messages are useful if we're crashing and displaying the error log, but since we'll be catching these I think it's more annoying then anything to just get the same message twice. The generic message at least tells you that it was a dictionary key-access issue which is "new information"

Kalobi replied:
Yeah that makes sense.

Kalobi commented on Celeste.Mod.mm/Patches/Level.cs:
```diff
@@ -206,13 +206,15 @@ class patch_Level : Level {
-                } else {
-                    patch_LevelEnter.ErrorMessage = Dialog.Get("postcard_levelloadfailed").Replace("((sid))", Session.Area.GetSID());
+                if (patch_LevelEnter.ErrorMessage == null) {
+                    if (e is ArgumentOutOfRangeException && e.StackTrace.Contains("get_DefaultSpawnPoint")) {
```
I complained about all the other `StackTrace.Contains`, so I might as well mention this one, though I'm not sure how best to address it. If we wanted to get rid of it, that would probably involve adding a bounds check in the `DefaultSpawnPoint` getter.

Kalobi reviewed (commented):
I didn't do any actual testing, I'm assuming you've done plenty of that, especially given the screenshots.

Kalobi commented on Celeste.Mod.mm/Content/Dialog/English.txt:
```diff
@@ -22,6 +22,11 @@
 	POSTCARD_LEVELGONE= 		{big}Thank you ((player))!{/big}{n}But our {#ff1144}((sid)){#} is on another mountain.
 	POSTCARD_LEVELLOADFAILED= 	{big}Sorry!{/big}{n}But {#ff1144}((sid)){#} has failed loading.{n}Please send your {#44adf7}log.txt{#}{n}to the map builder.
 	POSTCARD_LEVELNOSPAWN= 		{big}Oops!{/big}{n}Someone misplaced your {#44adf7}starting point!{#}{n}Please check if your map{n}contains a {#44adf7}Player{#} entity.
+	POSTCARD_BADTILEID=			{big}Oops!{/big}{n}There is an undefined ((type))tile id {#ff1144}((id)){#} at{n}{#ff1144}(((x)), ((y))){#} in room {#ff1144}((room)){#}{n}Check your {#44adf7}log.txt{#} for more info.
```
```suggestion
	POSTCARD_BADTILEID=			{big}Oops!{/big}{n}There is an undefined ((type))tile id {#ff1144}((id)){#} at{n}{#ff1144}(((x)), ((y))){#} in room {#ff1144}((room)){#}.{n}Check your {#44adf7}log.txt{#} for more info.
```

bigkahuna443 replied:
The lack of periods is actually intentional lol. You'll notice they're all after the variable names since those can potentially be very long. If they're too long, the period will print on the next line by itself which looks way worse than just not having a period after an SID. Will leave this open and mark the similar comments as resolved for now

Kalobi commented on Celeste.Mod.mm/Content/Dialog/English.txt:
```diff
@@ -22,6 +22,11 @@
 	POSTCARD_LEVELLOADFAILED= 	{big}Sorry!{/big}{n}But {#ff1144}((sid)){#} has failed loading.{n}Please send your {#44adf7}log.txt{#}{n}to the map builder.
 	POSTCARD_LEVELNOSPAWN= 		{big}Oops!{/big}{n}Someone misplaced your {#44adf7}starting point!{#}{n}Please check if your map{n}contains a {#44adf7}Player{#} entity.
+	POSTCARD_BADTILEID=			{big}Oops!{/big}{n}There is an undefined ((type))tile id {#ff1144}((id)){#} at{n}{#ff1144}(((x)), ((y))){#} in room {#ff1144}((room)){#}{n}Check your {#44adf7}log.txt{#} for more info.
+	POSTCARD_DECALREGERROR=		{big}Oops!{/big}{n}There was an error applying the{n}{#ff1144}((property)){#} property to {#ff1144}((decal)){#}{n}You may have a missing or invalid attribute in your {#44adf7}Decal Registry{#}.
```
```suggestion
	POSTCARD_DECALREGERROR=		{big}Oops!{/big}{n}There was an error applying the{n}{#ff1144}((property)){#} property to {#ff1144}((decal)){#}.{n}You may have a missing or invalid attribute in your {#44adf7}Decal Registry{#}.
```

Kalobi commented on Celeste.Mod.mm/Content/Dialog/English.txt:
```diff
@@ -22,6 +22,11 @@
+	POSTCARD_BADTILEID=			{big}Oops!{/big}{n}There is an undefined ((type))tile id {#ff1144}((id)){#} at{n}{#ff1144}(((x)), ((y))){#} in room {#ff1144}((room)){#}{n}Check your {#44adf7}log.txt{#} for more info.
+	POSTCARD_DECALREGERROR=		{big}Oops!{/big}{n}There was an error applying the{n}{#ff1144}((property)){#} property to {#ff1144}((decal)){#}{n}You may have a missing or invalid attribute in your {#44adf7}Decal Registry{#}.
+	POSTCARD_MISSINGTUTORIAL=	{big}Oops!{/big}{n}The playback tutorial {#ff1144}((tutorial)){#} could not be found.{n}Check your {#44adf7}log.txt{#} for more info.
+	POSTCARD_TILEXMLERROR=		{big}Oops!{/big}{n}There was an error parsing a tileset in {#ff1144}((path)){#}{n}Check your {#44adf7}log.txt{#} for more info.
```
```suggestion
	POSTCARD_TILEXMLERROR=		{big}Oops!{/big}{n}There was an error parsing a tileset in {#ff1144}((path)){#}.{n}Check your {#44adf7}log.txt{#} for more info.
```

Kalobi commented on Celeste.Mod.mm/Content/Dialog/English.txt:
```diff
@@ -22,6 +22,11 @@
 	POSTCARD_LEVELGONE= 		{big}Thank you ((player))!{/big}{n}But our {#ff1144}((sid)){#} is on another mountain.
 	POSTCARD_LEVELLOADFAILED= 	{big}Sorry!{/big}{n}But {#ff1144}((sid)){#} has failed loading.{n}Please send your {#44adf7}log.txt{#}{n}to the map builder.
 	POSTCARD_LEVELNOSPAWN= 		{big}Oops!{/big}{n}Someone misplaced your {#44adf7}starting point!{#}{n}Please check if your map{n}contains a {#44adf7}Player{#} entity.
```
While we're in here, should we change this to "your map's starting room" instead of "your map"?

Kalobi replied:
Also maybe `player` should be lowercase, since that's the actual name of the entity in the map data.

bigkahuna443 replied:
Personally I think it's pretty intuitive since the error occurs on loading into the first room. Plus I tried to avoid having to re-translate any of the old messages lol

bigkahuna443 committed:
Improve catch logic

1. Add ExceptionHelper file with some stacktrace search methods, use it to make catch logic more reliable.
2. Add AutotilerException, cleaner way to pass in the arguments than using the generic Data dictionary imo. On the fence on whether we should apply that the entire file or not?
3. Remove line from XML error :( it can be misleading depending on the error type
